### PR TITLE
test(console): integration tests for api resources

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

test(console): integration tests for api resources

## Testing

Integration tests passed.

### CODE REVIEW

File: packages/integration-tests/tests/api-resources.test.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+describe('admin console api resources', () => {
+  it('should get management api resource details successfully', async () => {
+    const fetchedManagementApiResource = await authedAdminApi
+      .get(`resources/${managementResource.id}`)
+      .json<Resource>();
+
+    expect(fetchedManagementApiResource.id).toBe(managementResource.id);
+    expect(fetchedManagementApiResource.name).toBe(managementResource.name);
+    expect(fetchedManagementApiResource.indicator).toBe(managementResource.indicator);
```
Comment (gao-sun) [inline packages/integration-tests/tests/api-resources.test.ts:20]: deepEqual? or contains?

File: packages/integration-tests/tests/api-resources.test.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+          indicator: testApiResource.indicator,
+        },
+      })
+      .json<Resource>();
+
+    expect(createdResource.name).toBe(testApiResource.name);
+    expect(createdResource.indicator).toBe(testApiResource.indicator);
+
+    // eslint-disable-next-line @silverhand/fp/no-mutation
+    testApiResource.id = createdResource.id;
```
Comment (gao-sun) [inline packages/integration-tests/tests/api-resources.test.ts:37]: you should not mutate a global var

File: packages/integration-tests/tests/api-resources.test.ts (hunk 1 above)
Comment (IceHe) [reply packages/integration-tests/tests/api-resources.test.ts:20]: better use `expect.objectContaining` to simplify. for example

image

File: packages/integration-tests/tests/api-resources.test.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    expect(updatedApiResource.name).toBe(newResourceName);
+    expect(updatedApiResource.accessTokenTtl).toBe(accessTokenTtl);
+  });
+
+  it('should delete api resource successfully', async () => {
+    expect(testApiResource.id).toBeTruthy();
+
+    await authedAdminApi.delete(`resources/${testApiResource.id}`);
+
+    const resources = await authedAdminApi.get('applications').json<Resource[]>();
```
Comment (simeng-li) [inline packages/integration-tests/tests/api-resources.test.ts:73]: 'applications'?

File: packages/integration-tests/tests/api-resources.test.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    expect(updatedResource.id).toBe(resource.id);
+    expect(updatedResource.name).toBe(newResourceName);
+    expect(updatedResource.accessTokenTtl).toBe(newAccessTokenTtl);
+  });
+
+  it('should delete api resource successfully', async () => {
+    const resourceName = 'bar';
+    const resourceIndicator = 'https://bar.logto.io';
+
+    const createdResource = await createResource(resourceName, resourceIndicator);
```
Comment (gao-sun) [inline packages/integration-tests/tests/api-resources.test.ts:71]: nit: this can be a helper function

**Review by gao-sun — APPROVED** [review]

**Review by gao-sun — CHANGES_REQUESTED** [review]
wait

**Review by gao-sun — APPROVED** [review]

File: packages/integration-tests/tests/api-resources.test.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+import { authedAdminApi } from '@/api';
+import { generateResourceIndicator, generateResourceName } from '@/utils';
+
+const createResource = (name?: string, indicator?: string) =>
+  authedAdminApi
+    .post('resources', {
+      json: {
+        name: name ?? generateResourceName(),
+        indicator: indicator ?? generateResourceIndicator(),
```
Comment (simeng-li) [inline packages/integration-tests/tests/api-resources.test.ts:12]: Do we have to generate a random resourceName & Indicator?  Or just manually gibe some static value is fine?

File: packages/integration-tests/tests/api-resources.test.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    const createdResource = await createResource();
+
+    expect(createdResource).toBeTruthy();
+
+    await authedAdminApi.delete(`resources/${createdResource.id}`);
+
+    const resources = await authedAdminApi.get('resources').json<Resource[]>();
+
+    const hasCreatedResource = resources.some((resource) => resource.id === createdResource.id);
+    expect(hasCreatedResource).toBeFalsy();
```
Comment (simeng-li) [inline packages/integration-tests/tests/api-resources.test.ts:75]: same here nit: use the resources/${id} API, and expect to throw would be more straightforward?

File: packages/integration-tests/tests/api-resources.test.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    expect(resources.some((resource) => resource.id === createdResource.id)).toBeTruthy();
+  });
+
+  it('should update api resource details successfully', async () => {
+    const resource = await createResource();
+
+    expect(resource).toBeTruthy();
+
+    const newResourceName = 'foo';
+    expect(resource.name).not.toBe(newResourceName);
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/api-resources.test.ts:46]: We want to check the resource name has been updated, so add this line to ensure the origin resource name is not the new resource name; then, we can know the name is updated with the new name.

File: packages/integration-tests/tests/api-resources.test.ts (hunk 5 above)
Comment (xiaoyijun) [reply packages/integration-tests/tests/api-resources.test.ts:12]: The indicator is unique, and I also want to keep creating resources with different names.

File: packages/integration-tests/src/utils.ts [hunk 8]
```diff
@@ -6,5 +6,10 @@ export const extractCookie = (response: Response) => {
   return headers['set-cookie']?.join('; ') ?? '';
 };
 
-export const generateUsername = () => `usr_${crypto.randomUUID().replaceAll('-', '_')}`;
-export const generatePassword = () => `pwd_${crypto.randomUUID()}`;
+const randomString = (length = 6) => crypto.randomUUID().slice(0, Math.max(0, length));
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/src/utils.ts]: It seems that this `randomString` method is useless, removed.

File: packages/integration-tests/tests/api-resources.test.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        indicator: indicator ?? generateResourceIndicator(),
+      },
+    })
+    .json<Resource>();
+
+describe('admin console api resources', () => {
+  it('should get management api resource details successfully', async () => {
+    const fetchedManagementApiResource = await authedAdminApi
+      .get(`resources/${managementResource.id}`)
+      .json<Resource>();
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/api-resources.test.ts:21]: Converting to `JSON` is not a typical case.
Sometimes we need to check the response status code without converting the response body to JSON.

**Review by simeng-li — APPROVED** [review]
LGTM

**Pull request merged**